### PR TITLE
[Broker] Avoid call sync method in async rest API for reset cursor

### PR DESCRIPTION
**Pull request by nodece**
Signed-off-by: Zixuan Liu <nodeces@gmail.com>

### Motivation

Avoid zk callback thread blocked, same with #13666

### Modifications

- Change sync method to async

### Documentation

Need to update docs? 

- [x] `no-need-doc`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 1]
```diff
@@ -1957,154 +1957,138 @@ private void internalExpireMessagesForAllSubscriptionsForNonPartitionedTopic(Asy
...
-            try {
-                validateGlobalNamespaceOwnership(namespaceName);
-            } catch (Exception e) {
-                log.warn("[{}][{}] Failed to reset cursor on subscription {} to time {}: {}",
-                        clientAppId(), topicName,
-                        subName, timestamp, e.getMessage());
-                resumeAsyncResponseExceptionally(asyncResponse, e);
-                return;
-            }
+            future = validateGlobalNamespaceOwnershipAsync(namespaceName);
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:1963]: if (topicName.isGlobal()) {
            future = validateGlobalNamespaceOwnershipAsync(namespaceName);
} else {
             future = CompletableFuture.completedFuture(null);
}

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 2]
```diff
@@ -1957,154 +1957,138 @@ private void internalExpireMessagesForAllSubscriptionsForNonPartitionedTopic(Asy
...
-        } else {
-            getPartitionedTopicMetadataAsync(topicName,
-                    authoritative, false).thenAccept(partitionMetadata -> {
-                final int numPartitions = partitionMetadata.partitions;
-                if (numPartitions > 0) {
-                    final CompletableFuture<Void> future = new CompletableFuture<>();
-                    final AtomicInteger count = new AtomicInteger(numPartitions);
-                    final AtomicInteger failureCount = new AtomicInteger(0);
-                    final AtomicReference<Throwable> partitionException = new AtomicReference<>();
+                    Throwable cause = (ex instanceof CompletionException ? ex.getCause() : ex);
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:1985]: FutureUtil.unwrapCompletionException()

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 3]
```diff
@@ -1957,154 +1957,138 @@ private void internalExpireMessagesForAllSubscriptionsForNonPartitionedTopic(Asy
...
+    private CompletableFuture<Void> internalResetCursorForPartitionedTopicAsync(String subName, long timestamp,
+                                                                                boolean authoritative) {
+        return getPartitionedTopicMetadataAsync(topicName, authoritative, false)
+                .thenCompose(partitionMetadata -> {
+                    final int numPartitions = partitionMetadata.partitions;
+                    if (numPartitions > 0) {
+                        final CompletableFuture<Void> result = new CompletableFuture<>();
+                        final AtomicInteger count = new AtomicInteger(numPartitions);
+                        final AtomicInteger failureCount = new AtomicInteger(0);
+                        final AtomicReference<Throwable> partitionException = new AtomicReference<>();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2012]: Please check here https://github.com/apache/pulsar/pull/13805/files#diff-66aeb65a64cbe7c541f013ae807c5bcbeab567bef77706c7ff2e0cbfe0d77eb1R623

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 4]
```diff
@@ -1957,154 +1957,138 @@ private void internalExpireMessagesForAllSubscriptionsForNonPartitionedTopic(Asy
...
-                return;
-            }
+            future = validateGlobalNamespaceOwnershipAsync(namespaceName);
         }
 
-        validateTopicOwnership(topicName, authoritative);
-        validateTopicOperation(topicName, TopicOperation.RESET_CURSOR, subName);
+        future.thenCompose(unused -> validateTopicOwnershipAsync(topicName, authoritative))
+                .thenCompose(unused -> validateTopicOperationAsync(topicName, TopicOperation.RESET_CURSOR, subName))
+                .thenCompose(unused -> {
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:1968]:
```suggestion
        future.thenCompose(__ -> validateTopicOwnershipAsync(topicName, authoritative))
                .thenCompose(__ -> validateTopicOperationAsync(topicName, TopicOperation.RESET_CURSOR, subName))
                .thenCompose(__ -> {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 5]
```diff
@@ -1957,154 +1958,135 @@ private void internalExpireMessagesForAllSubscriptionsForNonPartitionedTopic(Asy
...
-            });
-        }
+                });
     }
 
-    private void internalResetCursorForNonPartitionedTopic(AsyncResponse asyncResponse, String subName, long timestamp,
-                                       boolean authoritative) {
-        try {
-            validateTopicOwnership(topicName, authoritative);
-            validateTopicOperation(topicName, TopicOperation.RESET_CURSOR, subName);
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2065]: Why remove these two lines(2064~2065)?

Comment (nodece) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2065]: We have been checked at line 1969-1970.

**Pull request closed without merging**